service/s3/s3manager/upload.go

```diff
@@ -504,20 +520,67 @@ func (a completedParts) Len() int           { return len(a) }
 func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
 
+// request all the parts that are known to be completed for this upload
+func (u *multiuploader) requestCompletedParts() error {
+	var partNumberMarker *int64
+
+	for {
+		req := u.cfg.S3.ListPartsRequest(&s3.ListPartsInput{
+			Bucket:           u.uploader.in.Bucket,
+			Key:              u.uploader.in.Key,
+			UploadId:         &u.uploader.cfg.ResumeUploadID,
+			PartNumberMarker: partNumberMarker,
+		})
+
+		output, err := req.Send(u.ctx)
+		if err != nil {
+			return err
+		}
+
+		for _, part := range output.Parts {
+			completed := s3.CompletedPart{ETag: part.ETag, PartNumber: part.PartNumber}
+			u.parts = append(u.parts, completed)
+		}
+
+		if !*output.IsTruncated {
+			break
+		} else {
+			partNumberMarker = output.NextPartNumberMarker
+		}
+	}
+
+	return nil
+}
+
 // upload will perform a multipart upload using the firstBuf buffer containing
 // the first chunk of data.
 func (u *multiuploader) upload(firstBuf io.ReadSeeker) (*UploadOutput, error) {
```

Comment:
When a `ResumeUploadID` is provided, the uploader probably should not attempt to upload a single part object, i.e. PutObject. A `ResumeUploadID` implies that a multi part upload is being performed, and it may be unexpected to upload a single part object instead.

Reply:
Agreed.